classes/class-woothemes-sensei-certificates.php

```diff
@@ -1063,21 +1064,28 @@ public function certificate_link( $message, $course_id = 0 ) {
 			}
 		}
 
+		if ( empty( $user_id ) || 0 === $user_id ) {
+			$user_id = get_current_user_id();
+		}
+
 		$certificate_template_id = get_post_meta( $course_id, '_course_certificate_template', true );
 
 		if ( ! $certificate_template_id ) {
 			return $message;
 		}
 
+		$certificate_id     = $this->get_certificate_id( $course_id, $user_id );
 		$my_account_page_id = intval( Sensei()->settings->settings['my_course_page'] );
 		$view_link_courses  = Sensei()->settings->settings['certificates_view_courses'];
 		$view_link_profile  = Sensei()->settings->settings['certificates_view_profile'];
 		$is_viewable        = false;
 
-		if ( ( 'page' == get_post_type( $my_account_page_id )
+		if ( ( $this->can_view_certificate( $certificate_id ) )
```

Comment:
What do you think of moving the more performance-heavy checks after the "lighter" ones?

The `get_certificate_id` and `can_view_certificate` methods make database queries that can be avoided by doing all the lighter checks first and an early return if they fail.

```patch
diff --git a/classes/class-woothemes-sensei-certificates.php b/classes/class-woothemes-sensei-certificates.php
index 196457aa7..c6d703122 100644
--- a/classes/class-woothemes-sensei-certificates.php
+++ b/classes/class-woothemes-sensei-certificates.php
@@ -1074,14 +1074,12 @@ class WooThemes_Sensei_Certificates {
                        return $message;
                }

-               $certificate_id     = $this->get_certificate_id( $course_id, $user_id );
                $my_account_page_id = intval( Sensei()->settings->settings['my_course_page'] );
                $view_link_courses  = Sensei()->settings->settings['certificates_view_courses'];
                $view_link_profile  = Sensei()->settings->settings['certificates_view_profile'];
                $is_viewable        = false;

-               if ( ( $this->can_view_certificate( $certificate_id ) )
-                       && ( 'page' == get_post_type( $my_account_page_id )
+               if ( ( 'page' == get_post_type( $my_account_page_id )
                                || is_singular( 'course' )
                                || isset( $wp_query->query_vars['course_results'] )  && $view_link_courses
                                || isset( $wp_query->query_vars['learner_profile'] ) && $view_link_profile
@@ -1097,6 +1095,11 @@ class WooThemes_Sensei_Certificates {

                }

+               $certificate_id = $this->get_certificate_id( $course_id, $user_id );
+               if ( ! $this->can_view_certificate( $certificate_id ) ) {
+                       return $message;
+               }
+
                if ( is_singular( 'course' ) ) {

                        $certificate_url = $this->get_certificate_url( $post->ID, $user_id );
```

Reply:
Ah, that works great, thanks.

I'd originally set this up to use WP_Query but [had to switch to a db query](https://github.com/woocommerce/sensei-certificates/pull/341/commits/6d91ab1ff29a2b27ef2dba90a95dcd0eaf0b55cc) when I couldn't get the functionality I needed to work. Moving it down makes a lot of sense. Updated in f919929